src/Microsoft.Health.Fhir.Api/Features/Context/FhirAuthenticationExceptionHandlerMiddleware.cs

```diff
@@ -0,0 +1,48 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Health.Fhir.Api.Features.Context
+{
+    /// <summary>
+    /// Middlware that runs before authentication middleware and after FhirRequestContextBeforeAuthenticationMiddleware
+    /// </summary>
+    public class FhirAuthenticationExceptionHandlerMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public FhirAuthenticationExceptionHandlerMiddleware(
+            RequestDelegate next)
+        {
+            EnsureArg.IsNotNull(next, nameof(next));
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            try
+            {
+                await _next(context);
+            }
+            catch (Exception ex)
+            {
+                if (ex.InnerException is SecurityTokenException)
+                {
+                    ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
+                }
+                else
+                {
+                    throw;
```

Comment:
@brendankowitz I believe the intension is to use _ExceptionDispatchInfo.Capture(...).Throw()_ in both situations to keep the original call stack. 
The difference would be: if the inner exception is a SecurityTokenException, then we would throw the inner exception (using _ExceptionDispatchInfo.Capture(...).Throw()_ ), otherwise, throw the original exception (using _ExceptionDispatchInfo.Capture(...).Throw()_ ).